pkg/framework/provider/crypto/crypto.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+// Package crypto provides the default implementation of pkg/framework/spi/crypto.Crypto interface
+//
+// It uses github.com/tink/go crypto primitives
+package crypto
+
+import (
+	"errors"
+
+	"github.com/google/tink/go/aead"
+	"github.com/google/tink/go/core/primitiveset"
+	"github.com/google/tink/go/keyset"
+	"github.com/google/tink/go/signature"
+	aeadsubtle "github.com/google/tink/go/subtle/aead"
+	"golang.org/x/crypto/chacha20poly1305"
+)
+
+// Crypto is the default Crypto implementation
+type Crypto struct {
+}
+
+// Encrypt will encrypt msg using the implementation's corresponding encryption key and primitive in kh
+func (t *Crypto) Encrypt(msg, aad []byte, kh interface{}) ([]byte, []byte, error) {
+	keyHandle, ok := kh.(*keyset.Handle)
+	if !ok {
+		return nil, nil, errors.New("bad key handle format")
+	}
+
+	a, err := aead.New(keyHandle)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ct, err := a.Encrypt(msg, aad)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ps, err := keyHandle.Primitives()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ivSize := nonceSize(ps)
+	prefixLength := len(ps.Primary.Prefix)
+	cipherText := ct[prefixLength+ivSize:]
+	nonce := ct[prefixLength : prefixLength+ivSize]
+
+	return cipherText, nonce, nil
+}
+
+func nonceSize(ps *primitiveset.PrimitiveSet) int {
+	ivSize := aeadsubtle.AESGCMIVSize
+	// AESGCM and XChacha20Poly1305 nonce sizes supported only for now
+	if _, ok := (ps.Primary.Primitive).(*aeadsubtle.XChaCha20Poly1305); ok {
+		ivSize = chacha20poly1305.NonceSizeX
+	}
+
+	return ivSize
+}
+
+// Decrypt will decrypt cipher using the implementation's corresponding encryption key
+func (t *Crypto) Decrypt(cipher, nonce, aad []byte, kh interface{}) ([]byte, error) {
+	keyHandle, ok := kh.(*keyset.Handle)
+	if !ok {
+		return nil, errors.New("bad key handle format")
+	}
+
+	a, err := aead.New(keyHandle)
+	if err != nil {
+		return nil, err
+	}
+
+	ps, err := keyHandle.Primitives()
+	if err != nil {
+		return nil, err
+	}
+
+	ct := make([]byte, 0, len(ps.Primary.Prefix)+len(nonce)+len(cipher))
```

Comment:
a comment would be nice on why ct is being built this way.

Reply:
sure.. but this is how Tink assembles its cipher text for Decrypt() call.. this implementation is supposed to handle Tink specific format from the interface user where they only need to pass in the raw cipher text, nonce and aad used as is in the JWE envelope.

Reply:
done